lib/iris/tests/unit/fileformats/nc_load_rules/actions/__init__.py

```diff
@@ -0,0 +1,157 @@
+from iris.fileformats.cf import CFReader
+import iris.fileformats.netcdf
+from iris.fileformats.netcdf import _load_cube
+
+"""
+Notes on testing method.
+
+IN cf : "def _load_cube(engine, cf, cf_var, filename)"
+WHERE:
+  - engine is a :class:`iris.fileformats._nc_load_rules.engine.Engine`
+  - cf is a CFReader
+  - cf_var is a CFDAtaVariable
+
+As it's hard to construct a suitable CFReader from scratch, it would seem
+simpler (for now) to use an ACTUAL FILE.
+Likewise, the easiest approach to that is with CDL and "ncgen".
+To do this, we need a test "fixture" that can create suitable test files in a
+temporary directory.
+
+"""
+
+
+class Mixin__nc_load_actions:
+    """
+    Class to make testcases for rules or actions code, and check results.
+
+    Defines standard setUpClass/tearDownClass methods, to create a temporary
+    directory for intermediate files.
+    NOTE: owing to peculiarities of unittest, these must be explicitly called
+    from a setUpClass/tearDownClass within the 'final' inheritor, i.e. the
+    actual Test_XXX class which also inherits unittest.TestCase.
+
+    Testcases are manufactured by the '_make_testcase_cdl' method.
+    The 'run_testcase' method takes the '_make_testcase_cdl' kwargs and makes
+    a result cube (by: producing cdl, converting to netcdf, and loading the
+    'phenom' variable only).
+    Likewise, a generalised 'check_result' method will be used to perform result
+    checking.
+    Both '_make_testcase_cdl' and 'check_result' are not defined here :  They
+    are to be variously implemented by the inheritors.
+
+    """
+
+    # "global" test setting : whether to output various debug info
+    debug = False
+
+    @classmethod
+    def setUpClass(cls):
+        # Create a temp directory for temp files.
+        cls.temp_dirpath = Path(tempfile.mkdtemp())
+
+    @classmethod
+    def tearDownClass(cls):
+        # Destroy a temp directory for temp files.
+        shutil.rmtree(cls.temp_dirpath)
+
+    def load_cube_from_cdl(self, cdl_string, cdl_path, nc_path):
+        """
+        Load the 'phenom' data variable in a CDL testcase, as a cube.
+
+        Using ncgen, CFReader and the _load_cube call.
+
+        """
+        # Write the CDL to a file.
+        with open(cdl_path, "w") as f_out:
+            f_out.write(cdl_string)
+
+        # Create a netCDF file from the CDL file.
+        command = "ncgen -o {} {}".format(nc_path, cdl_path)
+        subprocess.check_call(command, shell=True)
+
+        # Simulate the inner part of the file reading process.
+        cf = CFReader(nc_path)
+        # Grab a data variable : FOR NOW always grab the 'phenom' variable.
+        cf_var = cf.cf_group.data_variables["phenom"]
+
+        engine = iris.fileformats.netcdf._actions_engine()
+
+        # If debug enabled, switch on the activation summary debug output.
+        # Use 'patch' so it is restored after the test.
+        self.patch("iris.fileformats.netcdf.DEBUG", self.debug)
```

Comment:
Where do we get `self.patch` from? Shouldn't the class have to derive from `tests.IrisTest` for this to work?

Reply:
Yes : the 'mixin' is only finally actually called from a 'test' class, which **_does_** inherit 'IrisTest'.
So, it's the same inheritance which provides all the 'assertXXX' methods.
(e.g. [here](https://github.com/SciTools/iris/blob/97ca13a532340a540e55cae701b985d8f758241a/lib/iris/tests/unit/fileformats/nc_load_rules/actions/test__hybrid_formulae.py#L21).
You don't see that in this file, because it is all in the inheritors.